### DISCUSSION TITLE
PascalABC.NET на ARM32 в сборке выпуска Windows 10 — упоминается на канале Discord/ RE: PascalABC.NET on ARM32 on Windows 10 release build - regarding claim on Discord channel

### DISCUSSION THREAD

**Discussion by c-j-womack** [pascalabcnet/pascalabcnet#2968] [Q&A]
Hi there,

I read on the Open RT discord channel that PascalABC.NET can run on ARM32 on leaked Windows 10 build.

The discord post stated
"If you want... to fully code on the RT and also build apps THERE and you don't mind using a diff language, take a look at PascalABC.Net, it's fully functional and you can create .exe files there, inside the RT and also aim for universal .exe, it works, it's purely Winforms and Pascal. "

However, the person who posted this no longer frequents that discord channel.

I'm wondering should I be able to use the existing binaries do they use anycpu?.... Or do I need to setup Visual Studio and compile for anycpu or ARM?

Unfortunately, it wasn't made clear in this or other posts whether he compiled the RT version from the source code.

Currently, I cannot run the current binaries on the Surface RT - I didn't include the PT4... 

I was able to run a binary generated on x86_64 machine after copying PABCRtl.dll. 

Sorry I don't speak russian {so I'll put a Google translation in)

Thanks or Спасибо

------------------------------------------

Всем привет,

Я прочитал на канале Discord Open RT, что PascalABC.NET может работать на ARM32 в просочившейся сборке Windows 10.

В сообщении разногласий говорилось
«Если вы хотите... полностью кодировать на RT, а также создавать приложения ТАМ, и вы не против использовать язык различий, взгляните на PascalABC.Net, он полностью функционален, и вы можете создавать файлы .exe там, внутри RT, а также нацелен на универсальный .exe, он работает, это чисто Winforms и Pascal».

Однако человек, опубликовавший это, больше не посещает этот канал Discord.

Мне интересно, смогу ли я использовать существующие двоичные файлы, они используют любой процессор?.... Или мне нужно настроить Visual Studio и скомпилировать для любого процессора или ARM?

К сожалению, ни в этом, ни в других постах не уточняется, скомпилировал ли он версию RT из исходного кода.

В настоящее время я не могу запустить текущие двоичные файлы на Surface RT — я не включил PT4...

Мне удалось запустить двоичный файл, созданный на компьютере x86_64, после копирования PABCRtl.dll.

Извините, я не говорю по-русски

**Comment by SunSerega** [comment]
I don't know any official Discord servers, the main one is currently on Telegram: https://t.me/PABCofficial

About translation: It's better to write only in English, or at least use something better than Google, like DeepL (and you'll still need to put in some effort cycling through translations for some words, but DeepL at least makes this process more intelligent). Google translated "discord post" as "post with lack of mutual understanding".

And consider using [discussions](https://github.com/pascalabcnet/pascalabcnet/discussions) or [forum](https://forum.mmcs.sfedu.ru/c/pascalabcnet/10) when you want to discuss something. But since you didn't know, that's already fine. Or, I can convert this issue to a discussion for you.

---

About the actual topic:

This Pascal, like any other .Net language (including C#), generates .exe files containing IL (Intermediate Language) code. Once the .exe file is executed, JIT (Just-In-Time) compiler takes the IL code and generated byte code (processor commands, like assembler but binary) optimized for the specific processor you run it on.

So all of the .Net .exe (and .dll) files are cross-platform by default, including the Pascal compiler and IDE (tho you can specify compiled files to only be 32-bit or 64-bit compatible).
But you need some kind of CLR (Common Language Runtime), containing a JIT compiler, garbage collector, etc. to run these .exe files.
On Windows, the CLR is .Net Framework (Or reworked .Net Core, but it still lacks some stuff the Pascal compiler needs to work)
On Linux, you either use Mono as a proper CLR or emulate Windows compatibility with something like Wine (and install .Net Framework just like you would on Windows).

ARM chips are just processors with fewer kinds of assembler commands, but in exchange, they are more efficient. Since JIT generated the instructions for the specific processor it runs on, as long as JIT (and the rest of CLR) itself can run - all the .Net programs would also work.
Unless the code isn't entirely managed (by CLR), like if it calls WinAPI functions, which are unmanaged because they are written in C/C++. Then you need system-level support. So in the case of WinAPI - it would run on Windows or from inside Wine on Linux, but you'll not get it to work natively in Linux.
But afaik, all of the Pascal IDE and Compiler work from within CLR, so this would only be the issue with your own programs written in Pascal.

**Comment by c-j-womack** [comment]
Hi there,

Feel free to convert this into discussion/forum instead. I'm sorry about using Google Translate instead of AI solution. I've seen people on other github repositories get annoyed when people post in another language.

I found you can use some old versions of PascalABC.NET on ARM32 if you extract the .exe file using 7-zip.

There's a discord server for Surface RT devices called "Open RT" and I subsequently found this post from 2021 "That would be great. In the meantime you might find PascalABC.Net a bit interesting. You can download it from the set of apps shared around here (or from the official website). It's Pascal/Delphi, it's an IDE, it runs, compiles and builds .exe's that run on the Surface RT WoA10, pretty clever, I've been playing a bit with it.

While you can also download it from the official website, the latest release didn't run on my side, but previous ones did." The person who posted this previously, his device died - hence he left that discord server "Open RT"

I found that newer versions of PascalABC.NET later then 14/03/2015 don't work on ARM32 (using the 7-zip extract trick) that were catalogued on Wayback machine didn't work (though haven't gone further to see if there's minimal work to fix)

[https://web.archive.org/web/20150315030956/http://pascalabc.net/downloads/PascalABCNETMiniSetup.exe](https://web.archive.org/web/20150315030956/http://pascalabc.net/downloads/PascalABCNETMiniSetup.exe)

I found if I use 7-zip to unpack the .exe file (which doesn't run on ARM32) I can then run PascalABCNET.exe inside.
However, there was no English translation in this so I had to revert back to 2015 in github (using [https://github.com/pascalabcnet/pascalabcnet/tree/HEAD@%7B2015-03-01%7D](https://github.com/pascalabcnet/pascalabcnet/tree/HEAD@%7B2015-03-01%7D) as mentioned in [https://stackoverflow.com/questions/21345787/github-repository-state-at-specified-time](https://stackoverflow.com/questions/21345787/github-repository-state-at-specified-time) look at the github /bin/Lng folder to download the English - https://github.com/pascalabcnet/pascalabcnet/tree/e6e67c193c17759ee72a42dce8c2e98784b1ca3a/bin/Lng and I used the GitZip chrome extension to do this. [https://chromewebstore.google.com/detail/gitzip-for-github/ffabmkklhbepgcgfonabamgnfafbdlkn?pli=1](https://chromewebstore.google.com/detail/gitzip-for-github/ffabmkklhbepgcgfonabamgnfafbdlkn?pli=1)
Pasting this folder inside the Lng folder enabled English translations once I changed in the GUI

I might investigate further the newer builds of PascalABC.NET or paste the error messages from these.
Thanks for your help.

**Reply by SunSerega** [reply]
> It's Pascal/Delphi

... no it's not. Delphi is a completely different dialect. And one of the only other Pascal dialects still surviving, but, in many ways, it's more archaic than PascalABC.Net.

> and I used the GitZip chrome extension to do this

Oh, I needed that the other day...

> extract the .exe file using 7-zip

Do you mean you are extracting the installer .exe file?
Btw, you can also download the [PABCNETC.ZIP](https://pascalabc.net/downloads/PABCNETC.zip), it's a portable version intended to be extracted that way. Though it only has the compiler and not IDE. And when extracting the installer - you are skipping some steps. But I guess it would be fine in most cases.
You can also try the Linux version - IDE from that package will likely have more chances to work. The compiler is the same in all packages tho.

> newer versions of PascalABC.NET later then 14/03/2015 don't work

"don't work" is never enough. I guess it crashes without proper error code/description if you just start the IDE or compiler .exe files. In that case, what if you start them from the console (some equivalent of Windows cmd)?

**Reply by cjwomack** [reply]
Hi there,

I was quoting someone's thread re about using it on ARM...  I quoted the whole bit including the wrong info re Language Spec not the just relevant bit -  sorry for starting bit of a flame war..

I know "don't work" isn't enough - I don't expect you guys to fix an esoteric case for an unsupported version of Windows on an unsupported architecture ARM32 unless I do some fairly serious troubleshooting and screenshots. That includes Google searches, discussions with ChatGPT etc, searches on Wayback machine if references to dead links.

I'm happy that at least one version is working,

**EDIT** Yes, the OS I'm running on the Surface RT, leaked win 10 build for ARM32, wouldn't let the installer.exe run... (I don't know why the installer .exe isn't architecture independent) hence I extracted the installer.exe after reading that post I quoted previously with the piece of incorrect info re PascalABC.NET then I thought about extracting the installer executable.

I wanted others to know what the current state of play is when it is fresh in my mind.

Yes, I download PascalABC.ZIP the other day, didn't try it yet.

Thanks for getting back to me.

**EDIT Oops used another browser with a different logged in account re my username**

**Reply by SunSerega** [reply]
> I don't know why the installer .exe isn't architecture independent

It's not written in .Net, it's created using NSIS - a tool used for most Windows installers. In particular, that allows it to automatically download and install the .Net Framework (if it's missing) before installing the actual PascalABC.Net.

> I wanted others to know what the current state of play is when it is fresh in my mind.

Alright, that's nice. But, for context, things are moving pretty quickly in this repo, so 2021 PascalABC.Net is also like a completely different language at this point, let alone 2015. Would be great to get the newest builds working too.
